src/lib.rs

```diff
@@ -0,0 +1,189 @@
+            Some(role) => role,
+            None => return PamResultCode::PAM_AUTH_ERR,
+        };
+        let conv = match pamh.get_item::<Conv>() {
+            Ok(Some(conv)) => conv,
+            Ok(None) => {
+                unreachable!("No conv available");
+            }
+            Err(err) => {
+                println!("Couldn't get pam_conv");
+                return err;
+            }
+        };
+        let state: &&str = &"unknown";
+        let rhost = match pamh.get_item::<RHost>() {
+            Ok(Some(rhost)) => rhost,
+            Ok(None) => {
+                unreachable!("No rhost available");
+            }
+            Err(err) => {
+                println!("Couldn't get pam_rhost");
+                return err;
+            }
+        };
+        let port = match get_free_tcp() {
+            Some(port) => port,
+            None => 0,
+        };
+        let authorize_url = format!("{}/oauth2/authorize?response_type=code&client_id={}&scope=guilds.members.read&state=fart&redirect_uri=http://{}:{}", API_ENDPOINT, client_id, rhost.to_str().unwrap(), port);    
+        pam_try!(conv.send(PAM_TEXT_INFO, &authorize_url));
+        let rt = Runtime::new().unwrap();
+        let gang = rt.block_on(generate_new_token(state, port));
+        let access_token = rt.block_on(get_access_token(client_id, client_secret, gang, format!("http://127.0.0.1:{}",port).to_string())).unwrap();
+        let actual_token_json = json::parse(&access_token).unwrap();
+        let actual_token = actual_token_json["access_token"].as_str().unwrap();
+        let member = rt.block_on(get_guild_member(actual_token, guild_id)).unwrap();
+        let member_json = json::parse(&member).unwrap();
+        if member_json["roles"].contains(role){
+            return PamResultCode::PAM_SUCCESS
+        }
+        pam_try!(conv.send(PAM_ERROR_MSG, "You do not have the required role!"));
+        PamResultCode::PAM_PERM_DENIED
+    }
+
+    fn sm_setcred(_pamh: &mut PamHandle, _args: Vec<&CStr>, _flags: PamFlag) -> PamResultCode {
+        println!("set credentials");
+        PamResultCode::PAM_SUCCESS
+    }
+
+    fn acct_mgmt(_pamh: &mut PamHandle, _args: Vec<&CStr>, _flags: PamFlag) -> PamResultCode {
+        println!("account management");
+        PamResultCode::PAM_SUCCESS
+    }
+}
+
+async fn generate_new_token(state: &str, port: u16) -> String {
+    let addr = ([0, 0, 0, 0], port).into();
+
+    let notify = Arc::new(Notify::new());
+    let token = Arc::new(Mutex::new(None::<String>));
+    let make_svc = make_service_fn({
+        let notify = notify.clone();
+        let token = token.clone();
+        move |_| {
+            let notify = notify.clone();
+            let token = token.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
+                    let notify = notify.clone();
+                    let token = token.clone();
```

Comment:
This is too much indentation. Surely this cannot be the best way to do this. Idk how to do it better tho because idk what you're trying to do <3